File: common/src/commonMain/kotlin/com/finance_tracker/finance_tracker/core/ui/CoinWidget.kt
package com.finance_tracker.finance_tracker.core.ui

import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Icon
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.finance_tracker.finance_tracker.core.common.`if`
import com.finance_tracker.finance_tracker.core.theme.CoinTheme

private val HorizontalPadding = 16.dp

@Composable
fun CoinWidget(
    title: String,
    modifier: Modifier = Modifier,
    withBorder: Boolean = false,
    withHorizontalPadding: Boolean = true,
    onClick: (() -> Unit)? = null,
    content: @Composable () -> Unit = {}
) {
    Column(
        modifier = modifier
            .`if`(onClick != null) {
                clickable { onClick?.invoke() }
            }
            .padding(vertical = 12.dp)
    ) {
        Row(
            modifier = Modifier
                .padding(horizontal = HorizontalPadding),
            horizontalArrangement = Arrangement.Start
        ) {
            Text(
                text = title,
                style = CoinTheme.typography.h5
            )
            if (onClick != null) {
                Spacer(
                    modifier = Modifier.weight(1f)
                )
                Icon(
                    painter = rememberVectorPainter("ic_arrow_next_small"),
                    contentDescription = null,
                    modifier = Modifier
                        .size(24.dp)
                        .align(Alignment.CenterVertically)
                )
            }
        }

        Box(
            modifier = Modifier
                .`if`(withHorizontalPadding) {
                    padding(horizontal = HorizontalPadding)
                }
                .padding(top = 12.dp)
                .`if`(withBorder) {
                    border(
                        width = 1.dp,
                        color = CoinTheme.color.dividers,
                        shape = RoundedCornerShape(12.dp)
                    )
                }
        ) {
            content.invoke()
        }
    }
}

File: ios/OneCoin/Core/Views/CoinWidget.swift
//
//  CoinWidget.swift
//  OneCoin
//
//  Created by Виталий Перятин on 22.04.2023.
//

import SwiftUI
import OneCoinShared

private let HorizontalPadding = 16.0

struct CoinWidget<Content: View>: View {
    
    @EnvironmentObject var theme: CoinTheme
    
    var title: String
    var withBorder: Bool = false
    var withHorizontalPadding: Bool = true
    var onClick: (() -> ())? = nil
    var content: () -> Content
    
    var body: some View {
        VStack(alignment: .leading) {
            CoinWidgetTitle(title: title, onClick: onClick)
            CoinWidgetContent<Content>(
                withBorder: withBorder,
                withHorizontalPadding: withHorizontalPadding,
                content: content
            )
        }
    }
}

private struct CoinWidgetTitle: View {
    
    @EnvironmentObject var theme: CoinTheme
    
    var title: String
    var onClick: (() -> ())? = nil
    
    var body: some View {
        let clickEnabled = onClick != nil
        HStack(alignment: .center) {
            Text(title)
                .fontH5Style(color: theme.colors.content)
            if (clickEnabled) {
                Spacer()
                // TODO: replace ic_arrow_down to ic_arrow_next_small
                SVGImageView(url: MR.files().ic_arrow_down.url)
                    .tintColor(theme.colors.content)
                    .frameSvg(width: 24, height: 24)
            }
        }
            .onTapGestureIf(clickEnabled) {
                onClick!()
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.vertical, 8)
            .padding(.horizontal, HorizontalPadding)
    }
}

private struct CoinWidgetContent<Content: View>: View {
    
    @EnvironmentObject var theme: CoinTheme
    
    var withBorder: Bool = false
    var withHorizontalPadding: Bool = true
    var content: () -> Content
    
    var body: some View {
        let contentHorizontalPadding = withHorizontalPadding ? HorizontalPadding : 0
        let borderWidth: CGFloat = withBorder ? 1 : 0
        let borderRadius: CGFloat = withBorder ? 12 : 0
        let shape = withBorder ? RoundedRectangle(cornerRadius: borderRadius) : RoundedRectangle(cornerRadius: 0)
        
        content()
            .frame(maxWidth: .infinity, alignment: .leading)
            .overlay(
                shape.stroke(theme.colors.dividers, lineWidth: borderWidth)
            )
            .background(theme.colors.background)
            .clipShape(shape)
            .padding(.horizontal, contentHorizontalPadding)
            .padding(.top, 4)
    }
}


struct CoinWidget_Previews: PreviewProvider {
    static var previews: some View {
        CoinWidget(
            title: "Title",
            withBorder: true,
            withHorizontalPadding: true
        ) {
            Text("Content")
        }
        .environmentObject(CoinTheme.light)
    }
}
